leshan-core/src/main/java/org/eclipse/leshan/core/node/TimestampedLwM2mNodes.java

```diff
@@ -0,0 +1,50 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Orange.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *     Orange - Send with multiple-timestamped values
+ *******************************************************************************/
+package org.eclipse.leshan.core.node;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A container for path {@link LwM2mPath} - nodes {@link LwM2mNode} map with optional timestamp segregation.
+ *
+ * @see TimestampedLwM2mNodesImpl
+ */
+public interface TimestampedLwM2mNodes {
+
+    /**
+     * Get timestamp grouped map of {@link LwM2mPath}-{@link LwM2mNode} map.
+     */
+    Map<Long, Map<LwM2mPath, LwM2mNode>> getTimestampedNodes();
+
+    /**
+     * Get all collected nodes represented as {@link LwM2mPath}-{@link LwM2mNode} map.
+     */
+    Map<LwM2mPath, LwM2mNode> getNodes();
```

Comment:
The TimestampedLwM2mNodes is collection of nodes with information about path and timestamp. This methods just returns nodes (path map). Should be called getAllNodes()? I'm not quite understand the problem.

Reply:
My concern is that I'm not sure what means "all nodes".

Eg. : 
```
t1 => path1 => nodeValue1a
t1 => path2 => nodeValue2
t2 => path1 => nodeValue1b
t3 => path1 => nodeValue1c

```
In this case,  with a map from path to node, we will have only 1 value for path1 or I missed something ? so this is not really all node ? or ?

Reply:
My intention in this case is provide way to get nodes as usual, ignoring timestamps (if you don't care about timestamps). In other case you can use getTimestampedNodes(), getNodesForTimestamp() to be time-aware. What name should I use to improve readability or give more context? And maybe javadoc is not enough decriptive?

Reply:
Ok I get it.
So I guess if you don't care about timestamped you want to get more recent value right ? So maybe this should be added in javadoc and/or the method is renamed in getMostRecentNodes() (or something like this) ? 

Bonus quesiton : 
Is it possible to have timestamped value mixed with "not timestamped" value ?  
If it is possible, null timestamped value is more or less recent ?

Reply:
I want to avoid the "most recent" phrase because it suggest to gets only ```"t3 => path1 => nodeValue1c"``` value from example above (because it's really most recent). My intention is to ignore timestamp segregation. Of course in example above we have conflicts and we have to choose one of "nodeValue1a", "nodeValue1b", "nodeValue1c" and the last one seems to be most appropriate indeed.

>  Is it possible to have timestamped value mixed with "not timestamped" value ?

If client send this kind of data and it's technically valid, the server should be able to handle it. So the answer is "yes"

> If it is possible, null timestamped value is more or less recent ?

It's undefined by definition, and until you use methods ```getTimestampedNodes()```, ```getNodesForTimestamp()``` and ```getTimestamps()``` it doesn't really matter (the consumer implementation have ability to manage the order). I had to provide some order for technical reasons (default TreeMap doesn't support null keys), so I recreated Comparator.nullsFirst() equivalent from java 8 (it's not copy paste!).

Maybe I should add some tests to valid order in cases pointed above?

Reply:
> I want to avoid the "most recent" phrase because it suggest to gets only "t3 => path1 => nodeValue1c" value from example above (because it's really most recent). My intention is to ignore timestamp segregation. Of course in example above we have conflicts and we have to choose one of "nodeValue1a", "nodeValue1b", "nodeValue1c" and the last one seems to be most appropriate indeed.

I understand but this seems more confusing to me to not explained how returned nodes are selected.
Even if the wording "most recent" is not crystal clear it gives at least some hints about the behavior which is only most recent value of each node is selected. 

Anyway, not a big deal if the method name does not contains the "most recent" idea but I think at least the javadoc should clearly define the behavior.

> It's undefined by definition, and until you use methods getTimestampedNodes(), getNodesForTimestamp() and getTimestamps() it doesn't really matter (the consumer implementation have ability to manage the order). I had to provide some order for technical reasons (default TreeMap doesn't support null keys), so I recreated Comparator.nullsFirst() equivalent from java 8 (it's not copy paste!).

I'm a bit confused by the idea because I don't know how this could be undefined.
At implementation level we must decide which node we keep in this path=>node map when we have something like : 
E.g. : 
```
null => path 1 => nodeValue1a
t1 => path 1 => nodeValue1b
t2 => path 1 => nodeValue1c
```
I guess `null` can be consider as now and so probably the most recent one and timestamp is for past value. (future value does not really make sense :thinking: )

> Maybe I should add some tests to valid order in cases pointed above?

It sounds a good idea.

Reply:
> I'm a bit confused by the idea because I don't know how this could be undefined.
I mean this order is nowhere directly described, so we have to establish how order null timestamp.
> I guess null can be consider as now and so probably the most recent one and timestamp is for past value.
Ok for me. I've updated implementation to take null timestamp as the latest.